File: model/src/main/java/com/bnorm/barkeep/model/Recipe.java
package com.bnorm.barkeep.model;

import java.util.Set;

import javax.annotation.Nullable;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonView;

@JsonInclude(JsonInclude.Include.NON_EMPTY)
public interface Recipe extends HasId, Comparable<Recipe> {

  @JsonView({Recipe.class, Book.class})
  @Nullable
  String getTitle();

  @JsonView({Recipe.class, Book.class})
  @Nullable
  String getDescription();

  @Nullable
  User getOwner();

  @JsonView({Recipe.class, Book.class})
  @Nullable
  String getImageUrl();

  @JsonView(Recipe.class)
  @Nullable
  String getInstructions();

  @JsonView(Recipe.class)
  @Nullable
  String getSource();

  @JsonView(Recipe.class)
  Set<Component> getComponents();

  @Override
  default int compareTo(Recipe o) {
    return COMPARATOR.compare(this, o);
  }
}

File: model/src/main/kotlin/com/bnorm/barkeep/model/Recipe.kt
// Copyright 2017 (C) BNORM Software. All rights reserved.
package com.bnorm.barkeep.model

import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonView

interface RecipeSpec {

  @get:JsonView(Recipe::class, Book::class)
  val title: String?

  @get:JsonView(Recipe::class, Book::class)
  val description: String?

  @get:JsonView(Recipe::class)
  val owner: User?

  @get:JsonView(Recipe::class, Book::class)
  val imageUrl: String?

  @get:JsonView(Recipe::class)
  val instructions: String?

  @get:JsonView(Recipe::class)
  val source: String?

  @get:JsonView(Recipe::class)
  val components: Set<Component>?
}

@JsonInclude(JsonInclude.Include.NON_EMPTY)
interface Recipe : RecipeSpec, HasId, Comparable<Recipe> {

  override fun compareTo(other: Recipe): Int {
    return HasId.COMPARATOR.compare(this, other)
  }
}
